### PR TITLE
(#638) Added requirements for etag handling

### PR DESCRIPTION
**Pull request by adrianhall**

### CODE REVIEW

File: docs/general/design.md [hunk 1]
```diff
@@ -130,6 +130,43 @@ Client libraries may support providing credential data via a connection string _
 
 {% include requirement/MUSTNOT id="general-auth-connection-strings" %} support constructing a service client with a connection string unless such connection string is available within tooling (for copy/paste operations).
 
+## Conditional requests
```
Comment (JeffreyRichter) [inline docs/general/design.md]: I'm confused by this guideline. There are services (like blobs) where this guideline will not work. So, is this guideline an option? (it can't be mandatory).
Also, there are some service (again Blobs), where multiple conditions with multiple Etag/Date values can be specified.

**Comment by adrianhall** [comment]
Updated this PR to include the comments from @annelo-msft

File: docs/general/design.md [hunk 2]
```diff
@@ -174,6 +174,65 @@ For methods that combine multiple requests into a single call:
 
 Such usage can cause confusion and will inevitably have to be changed on a per-language basis, which can cause consistency problems.
 
+## Conditional requests
+
+[Conditional requests](https://developer.mozilla.org/en-US/docs/Web/HTTP/Conditional_requests) are normally performed using HTTP headers that match the `ETag` to some known value.  For example, adding the following header will translate to "if the record has not been modified since the specified version".
```
Comment (annelo-msft) [inline docs/general/design.md]: It could also match a datetime in the If-Modified-Since or If-Unmodified-Since headers.

File: docs/general/design.md [hunk 3]
```diff
@@ -174,6 +174,65 @@ For methods that combine multiple requests into a single call:
...
+
```
Comment (annelo-msft) [inline docs/general/design.md]: Strictly speaking, this is correct, but since there is a different header called If-Modified-Since, I'd use language similar to "if the record's version, specified by the ETag, is not the same"

File: docs/general/design.md [hunk 4]
```diff
@@ -174,6 +174,65 @@ For methods that combine multiple requests into a single call:
...
+Not all services support all of these semantics, and may not support any of them.  Developers have varying levels of understanding of the `ETag` and conditional requests, so it is best to abstract this concept from the API surface.  There are two types of conditional requests we need to be concerned with:
+
+**Safe conditional requests** (e.g. GET)
+
+These are typically used to save bandwidth in an "update cache" scenario, i.e. I have a cached value, only send me the data if what the service has is newer than my copy. These return either a 200 or a 304 status code, indicating the value was not modified, which tells the caller that their cached value is up to date.
+
+**Unsafe conditional requests** (e.g. POST, PUT, or DELETE)
+
+These are typically used to prevent losing updates in an optimistic concurrency scenario, i.e. I've modified the cached value I'm holding, but don't update the service version unless it hass the same copy I've got. These return either a success or a 412 error status code, indicating the value was modified, to indicate to the caller that they'll need to retry their update if they want it to succeed.
```
Comment (annelo-msft) [inline docs/general/design.md]: typo on hass

File: docs/general/design.md [hunk 5]
```diff
@@ -174,6 +174,65 @@ For methods that combine multiple requests into a single call:
...
+These two cases are handled differently in client libraries.  However, the form of the call is the same in both cases.  The signature of the method should be:
+
+{% highlight text %}
+client.<method>(<item>, conditions)
+{% endhighlight %}
+
+For example, it is common to use a conditional get operation to handle caching situations (get the item, but only if not modified since the last time a synchronization was performed).  In this case, you might write the following:
+
+{% highlight javascript %}
+var item = await client.getItem(oldItem, conditions = MatchConditions.IfNotModified)
```
Comment (annelo-msft) [inline docs/general/design.md]: I had understood that we would not have a MatchConditions type as discussed in the first ETag meeting, but do what I described in the issue, which was the outcome of a later discussion with a subgroup of architects.  This involves passing an HttpRequestOptions type.

File: docs/general/design.md [hunk 6]
```diff
@@ -174,6 +174,65 @@ For methods that combine multiple requests into a single call:
...
+{% endhighlight %}
+
+Here, the `oldItem` model contains a reference to the `ETag` value.  The form of the method will be modified based on idiomatic usage patterns in the language of choice.  In cases where the `ETag` value is not known, the operation cannot be conditional.
+
+In addition to the `MatchConditions` enumerated type, the library developer can add a boolean operator that is set to `true` to establish the condition.  For example, use one of the following boolean names instead of the `conditions` operator:
+
+* `onlyIfMatches`
```
Comment (annelo-msft) [inline docs/general/design.md]: In my PR, I called these onlyIfChanged, and onlyIfUnchanged.

File: docs/general/design.md [hunk 7]
```diff
@@ -174,6 +174,65 @@ For methods that combine multiple requests into a single call:
...
+
+In addition to the `MatchConditions` enumerated type, the library developer can add a boolean operator that is set to `true` to establish the condition.  For example, use one of the following boolean names instead of the `conditions` operator:
+
+* `onlyIfMatches` 
+* `onlyIfNotPresent`
+* `onlyIfPresent`
+
+In all cases, the conditional expression is "opt-in".
+
+The return value from a conditional operation must be carefully considered.  For safe operators (e.g. GET), return the original value as you normally would from a non-conditional operation if a `Not Modified` result is received.  For unsafe operators (e.g. PUT, DELETE, or POST), throw a specific error when a `Precondition Failed` or `Conflict` result is received.  This allows the consumer to do something different in the case of conflicting results.
```
Comment (annelo-msft) [inline docs/general/design.md]: We decided that conditional methods would not throw for safe operations, but that referencing the value would throw, since by definition there is no value in the body to reference.

File: docs/general/design.md [hunk 8]
```diff
@@ -174,6 +174,65 @@ For methods that combine multiple requests into a single call:
...
+
+{% include requirement/SHOULD %} accept a `conditions` parameters (which takes an enumerated type) on service methods that allow a conditional check on the service. 
+
+{% include requirement/MUST %} include the `ETag` field as part of the object model when conditional operations are supported.
+
+{% include requirement/SHOULD %} accept additional boolean parameters on service methods as necessary to enable conditional checks on the service.
+
+{% include requirement/SHOULD %} return the original object when a `304 Not Modified` response is received from the service due to a conditional check.
```
Comment (annelo-msft) [inline docs/general/design.md]: We decided against this, see previous comment about throwing instead.

File: docs/general/design.md (hunk 2 above)
Comment (adrianhall) [reply docs/general/design.md]: I think I cover this in the headers list (and I also clarified)

File: docs/general/design.md (hunk 5 above)
Comment (adrianhall) [reply docs/general/design.md]: Got it - I'll adjust.

File: docs/general/design.md [hunk 9]
```diff
@@ -197,39 +197,34 @@ These are typically used to save bandwidth in an "update cache" scenario, i.e. I
...
 
 {% include requirement/SHOULD %} accept a `conditions` parameters (which takes an enumerated type) on service methods that allow a conditional check on the service. 
 
 {% include requirement/MUST %} include the `ETag` field as part of the object model when conditional operations are supported.
 
 {% include requirement/SHOULD %} accept additional boolean parameters on service methods as necessary to enable conditional checks on the service.
 
-{% include requirement/SHOULD %} return the original object when a `304 Not Modified` response is received from the service due to a conditional check.
+{% include requirement/SHOULD %} return a `Response<T>` when a `304 Not Modified` response is received from the service due to a conditional check.  If the value of such a response is accessed, throw an error.
```
Comment (annelo-msft) [inline docs/general/design.md]: I have recently learned that other languages don't have Response<T> so this is a Java & .NET specific guideline.  TS can't throw on a property reference because they don't have properties and may add a ConditionSucceeded field to response types returned from conditional requests that users must check to know whether the Response has content.  Python will throw on a 304 because this is idiomatic to Python.

Comment (johanste) [reply docs/general/design.md]: If we (Python) are free to return `None` for a 304/NotModified and throw on a 404/NotFound (as opposed to returning `None` for a 404 response), I would be open to reconsidering the "throw on 304" suggestion.

File: docs/general/design.md [hunk 10]
```diff
@@ -174,6 +174,60 @@ For methods that combine multiple requests into a single call:
...
+
+{% highlight text %}
+If-Not-Match: "etag-value"
+{% endhighlight %}
+
+With headers, tests are possible for the following:
+
+* Unconditionally (no additional headers)
+* If (not) modified since a version (`If-Match` and `If-Not-Match`)
+* If (not) modified since a date (`If-Updated-Since` and `If-Not-Updated-Since`)
```
Comment (JeffreyRichter) [inline docs/general/design.md]: Usually, these are called "If-Modified-Since" and "If-Unmodified-Since" see https://developer.mozilla.org/en-US/docs/Web/HTTP/Headers/If-Modified-Since and https://developer.mozilla.org/en-US/docs/Web/HTTP/Headers/If-Unmodified-Since

File: docs/general/design.md [hunk 11]
```diff
@@ -174,6 +174,60 @@ For methods that combine multiple requests into a single call:
...
+
+These are typically used to prevent losing updates in an optimistic concurrency scenario, i.e. I've modified the cached value I'm holding, but don't update the service version unless it has the same copy I've got. These return either a success or a 412 error status code, indicating the value was modified, to indicate to the caller that they'll need to retry their update if they want it to succeed.
+
+These two cases are handled differently in client libraries.  However, the form of the call is the same in both cases.  The signature of the method should be:
+
+{% highlight text %}
+client.<method>(<item>, requestOptions)
+{% endhighlight %}
+
+The `requestOptions` field is a type `HttpRequestOptions` in typed languages that provides additional options (such as additional headers) to the HTTP requres.   Where possible, the `ETag` value will be retrieved from the item that is passed into the method.  The form of the method will be modified based on idiomatic usage patterns in the language of choice.  In cases where the `ETag` value is not known, the operation cannot be conditional.
```
Comment (JeffreyRichter) [inline docs/general/design.md]: typo: requres --> request
You say "Where possible...": For storage (and probably some other services), this guidelines is NOT possible and so the ETags/Dates will be passed into the method as arguments. There should be a guideline for when this current guideline is NOT possible indicating that the ETags/Dates will be passed in. I think the structure that has these conditionals should be define in Azure Core and we should agree on the structure name so it it is similar across languages.

File: docs/general/design.md [hunk 12]
```diff
@@ -174,6 +174,60 @@ For methods that combine multiple requests into a single call:
...
+
+The library developer can add a boolean operator that is set to `true` to establish the condition.  For example, use one of the following boolean names instead of the `conditions` operator:
```
Comment (JeffreyRichter) [inline docs/general/design.md]: By 'operator' do you mean a boolean parameter?

File: docs/general/design.md [hunk 13]
```diff
@@ -174,6 +174,60 @@ For methods that combine multiple requests into a single call:
...
+
+* `onlyIfChanged` 
+* `onlyIfUnchanged`
```
Comment (JeffreyRichter) [inline docs/general/design.md]: For onlyIf(Un)Changed, it's unclear if the function is using an ETAG or a DATE to compare in the case where the passed-in object might have both an ETAG and a timestamp. Maybe onlyIfMatch, onlyIfNoneMatch, onlyIfModified, onlyIfUnmodified?

File: docs/general/design.md [hunk 14]
```diff
@@ -174,6 +174,60 @@ For methods that combine multiple requests into a single call:
...
+* `onlyIfMissing`
+* `onlyIfPresent`
+
+In all cases, the conditional expression is "opt-in".
+
+The return value from a conditional operation must be carefully considered.  For safe operators (e.g. GET), return a `Response<T>` from the conditional request.  However, throw an error if the `Value` is accessed, since there is no value in the body to reference.  For unsafe operators (e.g. PUT, DELETE, or POST), throw a specific error when a `Precondition Failed` or `Conflict` result is received.  This allows the consumer to do something different in the case of conflicting results.
+
+{% include requirement/SHOULD %} accept a `conditions` parameters (which takes an enumerated type) on service methods that allow a conditional check on the service. 
+
+{% include requirement/MUST %} include the `ETag` field as part of the object model when conditional operations are supported.
```
Comment (JeffreyRichter) [inline docs/general/design.md]: This can't be done for blobs. For this service, a separate structure containing conditions and ETAG/DATE values must be passed as an argument.

File: docs/general/design.md [hunk 15]
```diff
@@ -174,6 +174,60 @@ For methods that combine multiple requests into a single call:
...
+* `onlyIfChanged` 
+* `onlyIfUnchanged`
+* `onlyIfMissing`
+* `onlyIfPresent`
+
+In all cases, the conditional expression is "opt-in".
+
+The return value from a conditional operation must be carefully considered.  For safe operators (e.g. GET), return a `Response<T>` from the conditional request.  However, throw an error if the `Value` is accessed, since there is no value in the body to reference.  For unsafe operators (e.g. PUT, DELETE, or POST), throw a specific error when a `Precondition Failed` or `Conflict` result is received.  This allows the consumer to do something different in the case of conflicting results.
+
+{% include requirement/SHOULD %} accept a `conditions` parameters (which takes an enumerated type) on service methods that allow a conditional check on the service.
```
Comment (JeffreyRichter) [inline docs/general/design.md]: typo: parameters --> parameter

File: docs/general/design.md [hunk 16]
```diff
@@ -174,6 +174,60 @@ For methods that combine multiple requests into a single call:
...
+
+In all cases, the conditional expression is "opt-in".
+
+The return value from a conditional operation must be carefully considered.  For safe operators (e.g. GET), return a `Response<T>` from the conditional request.  However, throw an error if the `Value` is accessed, since there is no value in the body to reference.  For unsafe operators (e.g. PUT, DELETE, or POST), throw a specific error when a `Precondition Failed` or `Conflict` result is received.  This allows the consumer to do something different in the case of conflicting results.
+
+{% include requirement/SHOULD %} accept a `conditions` parameters (which takes an enumerated type) on service methods that allow a conditional check on the service. 
+
+{% include requirement/MUST %} include the `ETag` field as part of the object model when conditional operations are supported.
+
+{% include requirement/SHOULD %} accept additional boolean parameters on service methods as necessary to enable conditional checks on the service.
```
Comment (JeffreyRichter) [inline docs/general/design.md]: I'd rephrase to indicate a single boolean parameter on a single service method. As written it could be interpreted as multiple boolean parameters on a single service method.

File: docs/general/design.md [hunk 17]
```diff
@@ -174,6 +174,60 @@ For methods that combine multiple requests into a single call:
...
+
+{% include requirement/SHOULD %} return a `Response<T>` when a `304 Not Modified` response is received from the service due to a conditional check.  If the value of such a response is accessed, throw an error.
```
Comment (JeffreyRichter) [inline docs/general/design.md]: This is very .NET-specific and can not be applied to some other languages.

Comment (adrianhall) [reply docs/general/design.md]: Agreed - I've turned it around to say "should not throw an error on 304"

Comment (JeffreyRichter) [reply docs/general/design.md]: But, there may not be a Value to obtain. So, maybe in .NET and Java, throwing on Value access is OK but maybe for JS/Python/Go/C it isn't.

Comment (annelo-msft) [reply docs/general/design.md]: I've been talking to the Java team about this.  We should update the guidance to say, on a 304, return the same thing you would return for a 204.  Java returns Response<null> on a 204, so that's what they'll do for 304 as well.

It sounds like JS may be throwing when you access the value as well, although, @richardpark-msft, is this consistent with 204 behavior?

**Comment by bterlson** [comment]
LGTM, but is this missing how to handle if-modified-since? Seems like that has to be passed in as a value in the options bag (in js parlance anyway).

**Comment by adrianhall** [comment]
I'm expecting that per-language guidance plus some design work on the service feature team.  One team may have a lastModified field in a model, whereas another may need it explicitly specified.

File: docs/general/design.md [hunk 18]
```diff
@@ -174,6 +174,60 @@ For methods that combine multiple requests into a single call:
...
+* Unconditionally (no additional headers)
+* If (not) modified since a version (`If-Match` and `If-Not-Match`)
+* If (not) modified since a date (`If-Modified-Since` and `If-Unmodified-Since`)
+* If (not) present (`If-Match` and `If-Not-Match` with a `ETag=*` value)
+
+Not all services support all of these semantics, and may not support any of them.  Developers have varying levels of understanding of the `ETag` and conditional requests, so it is best to abstract this concept from the API surface.  There are two types of conditional requests we need to be concerned with:
+
+**Safe conditional requests** (e.g. GET)
+
+These are typically used to save bandwidth in an "update cache" scenario, i.e. I have a cached value, only send me the data if what the service has is newer than my copy. These return either a 200 or a 304 status code, indicating the value was not modified, which tells the caller that their cached value is up to date.
```
Comment (richardpark-msft) [inline docs/general/design.md]: Should this be '204' (ie, 'No Content'), not 200?

Comment (adrianhall) [reply docs/general/design.md]: Nope - it's definitely 304 Not Modified

Comment (annelo-msft) [reply docs/general/design.md]: Nope - 200 is success and returns the value if it's changed.  We could reword this sentence to make it more clear.

File: docs/general/design.md [hunk 19]
```diff
@@ -174,6 +174,60 @@ For methods that combine multiple requests into a single call:
...
+The library developer can add a boolean parameter that is set to `true` to establish the condition.  For example, use one of the following boolean names instead of the `conditions` operator:
+
+* `onlyIfChanged` 
+* `onlyIfUnchanged`
+* `onlyIfMissing`
+* `onlyIfPresent`
+
+In all cases, the conditional expression is "opt-in".
+
+The return value from a conditional operation must be carefully considered.  For safe operators (e.g. GET), return a `Response<T>` from the conditional request.  However, throw an error if the `Value` is accessed, since there is no value in the body to reference.  For unsafe operators (e.g. PUT, DELETE, or POST), throw a specific error when a `Precondition Failed` or `Conflict` result is received.  This allows the consumer to do something different in the case of conflicting results.
```
Comment (annelo-msft) [inline docs/general/design.md]: This is still .NET-ey.  Per our conversation, it would make sense to say something like "return whatever is the language-specific convention for successful responses without content (e.g. 204)."

Comment (bterlson) [reply docs/general/design.md]: +1 to that, @richardpark-msft and I were interpreting it that way anyway I think 😀

File: docs/general/design.md [hunk 20]
```diff
@@ -174,6 +174,60 @@ For methods that combine multiple requests into a single call:
...
+
+These are typically used to prevent losing updates in an optimistic concurrency scenario, i.e. I've modified the cached value I'm holding, but don't update the service version unless it has the same copy I've got. These return either a success or a 412 error status code, indicating the value was modified, to indicate to the caller that they'll need to retry their update if they want it to succeed.
+
+These two cases are handled differently in client libraries.  However, the form of the call is the same in both cases.  The signature of the method should be:
+
+{% highlight text %}
+client.<method>(<item>, requestOptions)
+{% endhighlight %}
+
+The `requestOptions` field is a type `HttpRequestOptions` in typed languages that provides additional options (such as additional headers) to the HTTP request.  The `Etag` value will be retrieved from the item that is passed into the method where possible, and method arguments where not possible. The form of the method will be modified based on idiomatic usage patterns in the language of choice.  In cases where the `ETag` value is not known, the operation cannot be conditional.
```
Comment (annelo-msft) [inline docs/general/design.md]: As an FYI, we've had a request to change this back to MatchConditions.  This can be updated after that decision is closed.

File: docs/general/design.md [hunk 21]
```diff
@@ -174,6 +174,60 @@ For methods that combine multiple requests into a single call:
...
+
+The library developer can add a boolean parameter that is set to `true` to establish the condition.  For example, use one of the following boolean names instead of the `conditions` operator:
+
+* `onlyIfChanged` 
+* `onlyIfUnchanged`
+* `onlyIfMissing`
+* `onlyIfPresent`
```
Comment (annelo-msft) [inline docs/general/design.md]: This is similar to the Python naming, whereas .NET decided to go with IfExists/IfNotExists.  We may want to call out that these should be idiomatic to the language.

File: docs/general/design.md [hunk 22]
```diff
@@ -174,6 +174,60 @@ For methods that combine multiple requests into a single call:
...
+
+In all cases, the conditional expression is "opt-in".
+
+The return value from a conditional operation must be carefully considered.  For safe operators (e.g. GET), return a `Response<T>` from the conditional request.  However, throw an error if the `Value` is accessed, since there is no value in the body to reference.  For unsafe operators (e.g. PUT, DELETE, or POST), throw a specific error when a `Precondition Failed` or `Conflict` result is received.  This allows the consumer to do something different in the case of conflicting results.
+
+{% include requirement/SHOULD %} accept a `conditions` parameter (which takes an enumerated type) on service methods that allow a conditional check on the service.
```
Comment (annelo-msft) [inline docs/general/design.md]: .NET will take a ConditionalResponseOptions (name to be finalized) object, not an enum.  Python is taking an enum.

File: docs/general/design.md (hunk 18 above)
Comment (richardpark-msft) [reply docs/general/design.md]: We have two paths, right?

204/304 -> content not changed, response body is probably empty or ignorable
200 -> response has been returned (same for conditional/safe GET and normal GET)
